backend/python/tapa/autobridge.py

```diff
@@ -291,40 +298,38 @@ def _add_connectivity_constraints(
-  (2) update the dataflow graph to add edges between async_mmap instances and task instances
+  (1) use a dummy node to represent the DDR/HBM port
+  (2) update the dataflow graph to add edges between dummy nodes and the task instance that access the port
+  (3) TODO: adjust the AXI pipeline according to the floorplanning results
   """
-  # add floorplan constraints for async_mmap and mmap
+  # async_mmap should be floorplanned together with its neighbor task instance
   for arg in top_task.args[kernel_arg]:
-    # if async_mmap, only floorplans the async_mmap instances.
-    # connections between the async_mmap instance and the task instance are pipelined
     if arg.cat == Instance.Arg.Cat.ASYNC_MMAP:
-      _logger.debug(f'constrain async_mmap instance {arg.mmap_name} to {region}')
-      config['OptionalFloorplan'][region].append(
-          rtl.async_mmap_instance_name(arg.mmap_name))
-    # normal mmap, the axi module is combined with the IO task instance
-    else:
-      _logger.debug('%s is constrained to %s because of %s', arg.instance.name,
-                  region, arg.name)
-      config['OptionalFloorplan'][region].append(arg.instance.name)
-  
-  # update the graph to add the FIFO conenctions for async_mmap
-  # the FIFO between async_mmap and task instances are not defined by tapa::stream
-  # so we need to manually insert them to the graph
+      config['GroupingConstraints'].append([arg.instance.name, rtl.async_mmap_instance_name(arg.mmap_name)])
+
   for arg in top_task.args[kernel_arg]:
-    if arg.cat == Instance.Arg.Cat.ASYNC_MMAP:
-      arg_width = width_table[arg.name]
-      
-      # 5 FIFOs between the async_mmap and task instance
-      # wr_data, wr_addr, rd_data, rd_addr, resp
-      total_connection_width = arg_width + arg_width + ADDR_CHANNEL_DATA_WIDTH + ADDR_CHANNEL_DATA_WIDTH + RESP_CHANNEL_DATA_WIDTH
-      
-      config['Edges'][f'{arg.mmap_name}_ch'] = {
-        'produced_by': rtl.async_mmap_instance_name(arg.mmap_name),
-        'consumed_by': arg.instance.name,
-        'width': total_connection_width,
-        'depth': 2,
-      }
+    # total width of an AXI channel
+    # wr_data, wr_addr, rd_data, rd_addr, resp + other control signals (approximately 50)
+    arg_width = width_table[arg.name]
+    total_connection_width = arg_width + arg_width + ADDR_CHANNEL_DATA_WIDTH + ADDR_CHANNEL_DATA_WIDTH + RESP_CHANNEL_DATA_WIDTH + 50
+    dummy_name = f'{_DUMMY_OFF_CHIP_MEMORY_PORT_PREFIX}_{region}_{arg.instance.name}'
+
+    config['OptionalFloorplan'][region].append(dummy_name)
+    config['Vertices'][dummy_name] = _DUMMY_VERTEX_NAME
+    config['Edges'][f'{dummy_name}_ch'] = {
+      'produced_by': dummy_name,
+      'consumed_by': arg.instance.name,
+      'width': total_connection_width,
+      'depth': 2,
+    }
+
+  config['Area'][_DUMMY_VERTEX_NAME] = {
+    "BRAM": 0,
+    "DSP": 0,
+    "FF": 0,
+    "LUT": 0,
+    "URAM": 0
+  }
```

Comment:
Why do we need this dummy vertex if its area is 0?

Reply:
It represents the virtual DDR/HBM. With this dummy vertex, we could have an edge that represents the AXI pipeline.